gcp_variant_transforms/libs/bigquery_vcf_schema.py

```diff
@@ -203,22 +210,57 @@ def generate_schema_from_header_fields(header_fields, variant_merger=None,
   return schema
 
 
-def get_row_from_variant(variant, split_alternate_allele_info_fields=True):
-  """Returns a BigQuery row according to the schema from the given variant.
+def get_rows_from_variant(variant, split_alternate_allele_info_fields=True):
+  """Yields BigQuery rows according to the schema from the given variant.
+
+  There is a 10MB limit for each BigQuqery row, which can exceed by having
+  a large number of calls. This method may split up a row into multiple rows if
+  it exceeds 10MB.
 
   Args:
     variant (``Variant``): Variant to process.
     split_alternate_allele_info_fields (bool): If true, all INFO fields with
       `Number=A` (i.e. one value for each alternate allele) will be stored under
       the `alternate_bases` record. If false, they will be stored with the rest
       of the INFO fields.
-  Returns:
-    A dict representing BigQuery row from the given variant.
+  Yields:
+    A dict representing a BigQuery row from the given variant. The row may have
+    a subset of the calls if it exceeds the maximum allowed BigQuery row size.
   Raises:
     ValueError: If variant data is inconsistent or invalid.
   """
   # TODO: Add error checking here for cases where the schema defined
   # by the headers does not match actual records.
+  base_row = _get_base_row_from_variant(
+      variant, split_alternate_allele_info_fields)
+  base_row_size_in_bytes = len(json.dumps(base_row))
+  row_size_in_bytes = base_row_size_in_bytes
+  row = copy.deepcopy(base_row)  # Keep base_row intact.
+  for call in variant.calls:
+    call_record = {
+        ColumnKeyConstants.CALLS_NAME: _get_bigquery_sanitized_field(call.name),
+        ColumnKeyConstants.CALLS_PHASESET: call.phaseset,
+        ColumnKeyConstants.CALLS_GENOTYPE: [g for g in call.genotype or []]
+    }
+    for key, field in call.info.iteritems():
+      if field is None:
+        continue
+      call_record[_get_bigquery_sanitized_field_name(key)] = (
+          _get_bigquery_sanitized_field(field))
+    # Add a few bytes to account for surrounding characters when concatenating.
+    call_record_size_in_bytes = len(json.dumps(call_record)) + 5
+    if (row_size_in_bytes + call_record_size_in_bytes >=
+        _MAX_BIGQUERY_ROW_SIZE_BYTES):
+      yield row
+      row = copy.deepcopy(base_row)
+      row_size_in_bytes = base_row_size_in_bytes
+    row[ColumnKeyConstants.CALLS].append(call_record)
+    row_size_in_bytes += call_record_size_in_bytes
+  yield row
+
+
+def _get_base_row_from_variant(variant, split_alternate_allele_info_fields):
+  """A helper method for ``get_row_from_variant`` to get row without calls."""
```

Comment:
It could be nice to turn rows into an object. This method could be implemented as a constructor and it could store row size and have methods that abstract a lot of the behavior in `get_rows_from_variants`.

Reply:
I agree! Great idea! I left it as a TODO for now though as it would require changing a bunch of stuff to do it properly and I prefer to not bundle it with this change.